requests_cache/core.py

```diff
@@ -4,250 +4,182 @@
         backend: str = None,
-        expire_after: Union[int, float, timedelta] = None,
+        expire_after: ExpirationTime = None,
         allowable_codes: Iterable[int] = (200,),
         allowable_methods: Iterable['str'] = ('GET', 'HEAD'),
         filter_fn: Callable = None,
         old_data_on_error: bool = False,
-        **kwargs
+        **kwargs,
     ):
         self.cache = backends.create_backend(backend, cache_name, kwargs)
-        self._cache_name = cache_name
-
-        if expire_after is not None and not isinstance(expire_after, timedelta):
-            expire_after = timedelta(seconds=expire_after)
-        self._cache_expire_after = expire_after
-        self._request_expire_after = 'default'
+        self.allowable_codes = allowable_codes
+        self.allowable_methods = allowable_methods
+        self.filter_fn = filter_fn or (lambda r: True)
+        self.old_data_on_error = old_data_on_error
 
-        self._cache_allowable_codes = allowable_codes
-        self._cache_allowable_methods = allowable_methods
-        self._filter_fn = filter_fn or (lambda r: True)
-        self._return_old_data_on_error = old_data_on_error
-        self._is_cache_disabled = False
+        self._cache_name = cache_name
+        self._expire_after = expire_after
+        self._request_expire_after = None
+        self._disabled = False
 
         # Remove any requests-cache-specific kwargs before passing along to superclass
-        session_kwargs = {k: v for k, v in kwargs.items() if k not in BACKEND_KWARGS}
+        session_kwargs = {k: v for k, v in kwargs.items() if k not in backends.BACKEND_KWARGS}
         super().__init__(**session_kwargs)
 
-    def _determine_expiration_datetime(self, relative_to=None):
-        """Determines the absolute expiration datetime for a response.
-        Requires :attr:`self._cache_expire_after` and :attr:`self._request_expire_after` to be set.
-        See :meth:`request` for more information.
-
-        :param response: the response (potentially loaded from the cache)
-        :type response: requests.Response
-        :param relative_to: Parameter for easy unit testing to fix ``now``,
-                            defaults to ``datetime.now(timezone.utc)`` for normal use.
-        :type relative_to: Union[None, datetime.datetime]
-        :return: The absolute expiration date
-        :rtype: datetime.datetime
+    @property
+    def expire_after(self):
+        """Get either the per-session expiration, or per-request expiration, if set"""
+        return self.request_expire_after or self._expire_after
+
+    @expire_after.setter
+    def expire_after(self, value: ExpirationTime):
+        """Get either the per-session expiration, or per-request expiration, if set"""
+        self._expire_after = value
+
+    @property
+    def request_expire_after(self) -> Optional[ExpirationTime]:
+        """Property to immediately reset temporary per-request expiration after being accessed"""
```

Comment:
Having a getter removing the value might be a bit confusing and could potentially lead to subtle bugs in the future. Maybe providing a more explicit method such as `pop_request_expire_after` or similar (`pop` might have slightly wrong semantics in this case as well, as we don't have a stack, but it usually follows the `return what you removed` principle).

Reply:
Yeah, I'm not quite happy with this either! A `pop` method might be better.

As you probably noticed when writing your PR, it would be even better if we could send that info along with the request info from `request() -> send()`, but there doesn't seem to be an easy way to do that. I spent some time digging through `requests` code, and the relevant parts are:
* [Session.request()](https://github.com/psf/requests/blob/master/requests/sessions.py#L470)
* [PreparedRequest.prepare()](https://github.com/psf/requests/blob/master/requests/models.py#L310)
* [Session.send()](https://github.com/psf/requests/blob/master/requests/sessions.py#L626)

I'll think about this one a little more.

Reply:
> As you probably noticed when writing your PR
Yes, it's unfortunately very brittle one way or the other, which is why I was thinking if naming can at least avoid some wtf-moments. But this could be deferred to a future PR.

Reply:
I dug through `requests` code some more, and it would be possible to temporarily set `_request_expire_after` in request headers or params. This seems like too much of a hack, though.

Instead I wrapped it in a contextmanager:
https://github.com/reclosedev/requests-cache/blob/7215cbcb9fda53852e85b811fc4779016ac15ec4/requests_cache/core.py#L64-L69

Which is only used in one place: https://github.com/reclosedev/requests-cache/blob/7215cbcb9fda53852e85b811fc4779016ac15ec4/requests_cache/core.py#L105-L112

So I don't think there's any way that could lead to unexpected behavior now.